pkg/reconciler/source/config_watcher.go

```diff
@@ -20,143 +20,212 @@ import (
+	envMetricsCfg = "K_METRICS_CONFIG"
+	envTracingCfg = "K_TRACING_CONFIG"
 )
 
+// ConfigWatcher keeps track of logging, metrics and tracing configurations by
+// watching corresponding ConfigMaps.
 type ConfigWatcher struct {
 	logger *zap.Logger
 
 	component string
 
-	loggingConfig *pkgLogging.Config
-	metricsConfig *metrics.ExporterOptions
-	tracingCfg    *tracingconfig.Config
+	loggingCfg *pkglogging.Config
+	metricsCfg *metrics.ExporterOptions
+	tracingCfg *tracingconfig.Config
 }
 
-func StartWatchingSourceConfigurations(loggingContext context.Context, component string, cmw configmap.Watcher) *ConfigWatcher {
-	cw := ConfigWatcher{
-		logger:    logging.FromContext(loggingContext),
+// configWatcherOption is a function option for ConfigWatchers.
+type configWatcherOption func(*ConfigWatcher, configmap.Watcher)
+
+// WatchConfigurations returns a ConfigWatcher initialized with the given
+// options. If no option is passed, the ConfigWatcher observes ConfigMaps for
+// logging, metrics and tracing.
+func WatchConfigurations(loggingCtx context.Context, component string,
+	cmw configmap.Watcher, opts ...configWatcherOption) *ConfigWatcher {
+
+	cw := &ConfigWatcher{
+		logger:    logging.FromContext(loggingCtx),
 		component: component,
 	}
 
-	if dcmw, ok := cmw.(configmap.DefaultingWatcher); ok {
-		dcmw.WatchWithDefault(corev1.ConfigMap{
-			ObjectMeta: metav1.ObjectMeta{Name: pkgLogging.ConfigMapName()},
-			Data:       map[string]string{},
-		}, cw.UpdateFromLoggingConfigMap)
-		dcmw.WatchWithDefault(corev1.ConfigMap{
-			ObjectMeta: metav1.ObjectMeta{Name: metrics.ConfigMapName()},
-			Data:       map[string]string{},
-		}, cw.UpdateFromMetricsConfigMap)
-		dcmw.WatchWithDefault(corev1.ConfigMap{
-			ObjectMeta: metav1.ObjectMeta{Name: tracingconfig.ConfigName},
-			Data:       map[string]string{},
-		}, cw.UpdateFromTracingConfigMap)
+	if len(opts) == 0 {
+		WithLogging(cw, cmw)
+		WithMetrics(cw, cmw)
+		WithTracing(cw, cmw)
+
 	} else {
-		cmw.Watch(pkgLogging.ConfigMapName(), cw.UpdateFromLoggingConfigMap)
-		cmw.Watch(metrics.ConfigMapName(), cw.UpdateFromMetricsConfigMap)
-		cmw.Watch(tracingconfig.ConfigName, cw.UpdateFromTracingConfigMap)
+		for _, opt := range opts {
+			opt(cw, cmw)
+		}
 	}
 
-	return &cw
+	return cw
 }
 
-func (r *ConfigWatcher) UpdateFromLoggingConfigMap(cfg *corev1.ConfigMap) {
-	if cfg != nil {
-		delete(cfg.Data, "_example")
-
-		logcfg, err := pkgLogging.NewConfigFromConfigMap(cfg)
-		if err != nil {
-			r.logger.Warn("failed to create logging config from configmap", zap.String("cfg.Name", cfg.Name))
-			return
-		}
-		r.loggingConfig = logcfg
-		r.logger.Debug("Update from logging ConfigMap", zap.Any("ConfigMap", cfg))
-	}
+// StartWatchingSourceConfigurations runs WatchConfigurations with all configurations.
+// For backwards compatibility only.
+func StartWatchingSourceConfigurations(loggingCtx context.Context, component string, cmw configmap.Watcher) *ConfigWatcher {
```

Comment:
I found this to be quite verbose, but I didn't want to savagely rename the function since you already started using it in some places. Let me know if you think it's fine to switch to the shorter option or now.